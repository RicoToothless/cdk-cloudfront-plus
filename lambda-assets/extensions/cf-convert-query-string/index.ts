import { URLSearchParams } from 'url'
export async function lambdaHandler(event: any) {
    /**
     * Whitelisted request arguments that will be converted into additional headers
     * `NEEDED_KEYS` is the placeholder for `args` that will be bundled and injected by esbuild.
     * */
    const neededKeys = NEEDED_KEYS

    console.log(`${JSON.stringify(event, null, '\t')}`);

    let request = event.Records[0].cf.request;
    let querystring = request.querystring;

    let beforeHeaders = request.headers;
    let beforeQueryString = querystring;

    // Parse request querystring to get dictionary/json
    let params = new URLSearchParams(querystring);

    console.log(`Before processing, the content of headers:${JSON.stringify(beforeHeaders, null, '\t')}`);
    console.log(`Before processing, the content of querystring:${JSON.stringify(beforeQueryString, null, '\t')}`);
    if (params != null) {
        // Add headers according to a query string
        const capitalizeHeaderOrNot = true
        for (let key of neededKeys) {
            _add_header(key, params, request, capitalizeHeaderOrNot);
        }
    }
    let afterHeader = request.headers;
    let afterQueryString = request.querystring;
    console.log(`After processing, the content of headers:${JSON.stringify(afterHeader, null, '\t')}`);
    console.log(`After processing, the content of querystring:${JSON.stringify(afterQueryString, null, '\t')}`);

    console.log(JSON.stringify(request, null, '\t'));
    return request;
}


function capitalizeFirstLetterForEachTearm(term: string, delimiter: string = '-'): string {
    let components = term.split(delimiter);
    components.forEach((item, index) => components[index] = item.charAt(0).toUpperCase() + item.slice(1));
    let result = components.join(delimiter);
    return result;
}

function _add_header(headerName: string, params: any, request: any, capitalizeOrNot: boolean = false): void {
    const properHeader = (capitalizeOrNot == true) ? capitalizeFirstLetterForEachTearm('x-'.concat(headerName)) : headerName;
    request.headers['x-'.concat(headerName)] = [
        { "key": properHeader, "value": (params.get(headerName) === null) ? '' : params.get(headerName) }]
    params.delete(headerName);
}
