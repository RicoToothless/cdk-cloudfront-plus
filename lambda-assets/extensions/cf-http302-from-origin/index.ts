import { String } from 'aws-sdk/clients/cloudsearch';
import axios, { AxiosResponse } from 'axios';

export interface HttpResponse {
  readonly status: number;
  readonly statusDescription: string;
  readonly headers: { [key: string]: {[key: string]: String}[] },
  readonly body: string,
}

export async function HttpGet(url: string): Promise<AxiosResponse> {
  const client = axios.create();
  const response: AxiosResponse<any> = await client.get(url)
  return response
}

export async function handler(event: any) {
  const response = event.Records[0].cf.response;
  console.log('response: %j', response)
  if (response.status === '302' && response.headers['location'].length != 0) {
    const locationUrl = response.headers['location'][0].value;
    const newResponse = await HttpGet(locationUrl)
    const responseObject: HttpResponse = {
      headers: {
        'cache-control': [{
          key: 'Cache-Control',
          value: `max-age=0`,
        }],
      },
      status: newResponse.status,
      statusDescription: newResponse.statusText,
      body: newResponse.data,
    }
    return responseObject
  } else {
    return response;
  }
};


function cacheControl(responseObject: HttpResponse,  maxAge: string) {
  return Object.assign(responseObject, {
    headers: {
      'cache-control': [{
        key: 'Cache-Control',
        value: `max-age=${maxAge}`,
      }],
    }
  })
}
