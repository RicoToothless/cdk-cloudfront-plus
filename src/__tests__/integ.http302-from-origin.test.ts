import { SynthUtils, arrayWith } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';
import * as cf from '@aws-cdk/aws-cloudfront';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as elbv2 from '@aws-cdk/aws-elasticloadbalancingv2';
import * as cdk from '@aws-cdk/core';
import * as extensions from '../extensions';

test('check HTTP302FromOrigin setting', () => {
  // GIVEN
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'http302-demo');

  // WHEN
  new extensions.HTTP302FromOrigin(stack, 'http302');

  // WHEN
  // check http302FromOrigin SAR version
  expect(stack).toHaveResourceLike('AWS::Serverless::Application', {
    Location: {
      ApplicationId: 'arn:aws:serverlessrepo:us-east-1:418289889111:applications/http302-from-origin',
      SemanticVersion: '1.0.2',
    },
  });

  // check where to get lambda version
  expect(stack).toHaveResourceLike('AWS::Lambda::Version', {
    FunctionName: {
      'Fn::Select': arrayWith(
        {
          'Fn::Split': arrayWith(
            {
              'Fn::GetAtt': [
                'http30232E1FD3B',
                'Outputs.Http302Function',
              ],
            },
          ),
        },
      ),
    },
  });
});

test('check ALB & cloudfront setting', () => {
  // GIVEN
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'http302-demo');

  // create the cloudfront distribution with extension(s)
  const http302FromOrigin = new extensions.HTTP302FromOrigin(stack, 'http302');

  // WHEN
  // ALB
  const publicSubnet = {
    cidrMask: 20,
    name: 'Public',
    subnetType: ec2.SubnetType.PUBLIC,
  };

  const vpc = new ec2.Vpc(stack, 'Vpc', {
    natGateways: 0,
    subnetConfiguration: [publicSubnet],
  });

  const demoAlb = new elbv2.ApplicationLoadBalancer(stack, 'Alb302Redirect', {
    vpc: vpc,
    internetFacing: true,
  });

  demoAlb.connections.allowFrom(ec2.Peer.anyIpv4(), ec2.Port.tcp(80));
  demoAlb.addListener('Alb302RedirectRule', {
    port: 80,
    defaultAction: elbv2.ListenerAction.redirect({
      protocol: 'HTTP',
      host: 'neverssl.com',
      path: '/',
    }),
  });

  // cloudfront
  new cf.CloudFrontWebDistribution(stack, 'DemoCloudFront', {
    viewerProtocolPolicy: cf.ViewerProtocolPolicy.ALLOW_ALL,
    originConfigs: [
      {
        customOriginSource: {
          originProtocolPolicy: cf.OriginProtocolPolicy.HTTP_ONLY,
          domainName: demoAlb.loadBalancerDnsName,
        },
        behaviors: [{
          isDefaultBehavior: true,
          lambdaFunctionAssociations: [http302FromOrigin],
        }],
      },
    ],
  });

  // THEN
  expect(SynthUtils.synthesize(stack).template).toMatchSnapshot();

  // check ALB redirect
  expect(stack).toHaveResourceLike('AWS::ElasticLoadBalancingV2::Listener', {
    DefaultActions: [
      {
        RedirectConfig: {
          Host: 'neverssl.com',
          Path: '/',
          Protocol: 'HTTP',
          StatusCode: 'HTTP_302',
        },
        Type: 'redirect',
      },
    ],
    Port: 80,
    Protocol: 'HTTP',
  });

  // check cloudfront setting & lambda association
  expect(stack).toHaveResourceLike('AWS::CloudFront::Distribution', {
    DistributionConfig: {
      DefaultCacheBehavior: {
        LambdaFunctionAssociations: [
          {
            EventType: 'origin-response',
            LambdaFunctionARN: {
              Ref: 'LambdaVersionhttp30247FAA424',
            },
          },
        ],
        ViewerProtocolPolicy: 'allow-all',
      },
      Origins: [
        {
          CustomOriginConfig: {
            OriginProtocolPolicy: 'http-only',
          },
          DomainName: {
            'Fn::GetAtt': [
              'Alb302Redirect610ABF15',
              'DNSName',
            ],
          },
        },
      ],
    },
  });
});
