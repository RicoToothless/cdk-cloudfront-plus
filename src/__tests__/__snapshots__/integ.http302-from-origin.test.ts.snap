// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check ALB & cloudfront setting 1`] = `
Object {
  "Resources": Object {
    "Alb302Redirect610ABF15": Object {
      "DependsOn": Array [
        "VpcPublicSubnet1DefaultRoute3DA9E72A",
        "VpcPublicSubnet2DefaultRoute97F91067",
      ],
      "Properties": Object {
        "LoadBalancerAttributes": Array [
          Object {
            "Key": "deletion_protection.enabled",
            "Value": "false",
          },
        ],
        "Scheme": "internet-facing",
        "SecurityGroups": Array [
          Object {
            "Fn::GetAtt": Array [
              "Alb302RedirectSecurityGroupDF0A30D2",
              "GroupId",
            ],
          },
        ],
        "Subnets": Array [
          Object {
            "Ref": "VpcPublicSubnet1Subnet5C2D37C4",
          },
          Object {
            "Ref": "VpcPublicSubnet2Subnet691E08A3",
          },
        ],
        "Type": "application",
      },
      "Type": "AWS::ElasticLoadBalancingV2::LoadBalancer",
    },
    "Alb302RedirectAlb302RedirectRule473E60D9": Object {
      "Properties": Object {
        "DefaultActions": Array [
          Object {
            "RedirectConfig": Object {
              "Host": "neverssl.com",
              "Path": "/",
              "Protocol": "HTTP",
              "StatusCode": "HTTP_302",
            },
            "Type": "redirect",
          },
        ],
        "LoadBalancerArn": Object {
          "Ref": "Alb302Redirect610ABF15",
        },
        "Port": 80,
        "Protocol": "HTTP",
      },
      "Type": "AWS::ElasticLoadBalancingV2::Listener",
    },
    "Alb302RedirectSecurityGroupDF0A30D2": Object {
      "Properties": Object {
        "GroupDescription": "Automatically created Security Group for ELB http302demoAlb302Redirect0EDBCCB6",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "255.255.255.255/32",
            "Description": "Disallow all traffic",
            "FromPort": 252,
            "IpProtocol": "icmp",
            "ToPort": 86,
          },
        ],
        "SecurityGroupIngress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "from 0.0.0.0/0:80",
            "FromPort": 80,
            "IpProtocol": "tcp",
            "ToPort": 80,
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "DemoCloudFrontCFDistributionDD56E6FE": Object {
      "Properties": Object {
        "DistributionConfig": Object {
          "DefaultCacheBehavior": Object {
            "AllowedMethods": Array [
              "GET",
              "HEAD",
            ],
            "CachedMethods": Array [
              "GET",
              "HEAD",
            ],
            "Compress": true,
            "ForwardedValues": Object {
              "Cookies": Object {
                "Forward": "none",
              },
              "QueryString": false,
            },
            "LambdaFunctionAssociations": Array [
              Object {
                "EventType": "origin-response",
                "LambdaFunctionARN": Object {
                  "Ref": "LambdaVersionhttp30247FAA424",
                },
              },
            ],
            "TargetOriginId": "origin1",
            "ViewerProtocolPolicy": "allow-all",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": Array [
            Object {
              "ConnectionAttempts": 3,
              "ConnectionTimeout": 10,
              "CustomOriginConfig": Object {
                "HTTPPort": 80,
                "HTTPSPort": 443,
                "OriginKeepaliveTimeout": 5,
                "OriginProtocolPolicy": "http-only",
                "OriginReadTimeout": 30,
                "OriginSSLProtocols": Array [
                  "TLSv1.2",
                ],
              },
              "DomainName": Object {
                "Fn::GetAtt": Array [
                  "Alb302Redirect610ABF15",
                  "DNSName",
                ],
              },
              "Id": "origin1",
            },
          ],
          "PriceClass": "PriceClass_100",
          "ViewerCertificate": Object {
            "CloudFrontDefaultCertificate": true,
          },
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "LambdaVersionhttp30247FAA424": Object {
      "Properties": Object {
        "FunctionName": Object {
          "Fn::Select": Array [
            6,
            Object {
              "Fn::Split": Array [
                ":",
                Object {
                  "Fn::GetAtt": Array [
                    "http30232E1FD3B",
                    "Outputs.Http302Function",
                  ],
                },
              ],
            },
          ],
        },
      },
      "Type": "AWS::Lambda::Version",
    },
    "Vpc8378EB38": Object {
      "Properties": Object {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "http302-demo/Vpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "VpcIGWD7BA715C": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "http302-demo/Vpc",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "VpcPublicSubnet1DefaultRoute3DA9E72A": Object {
      "DependsOn": Array [
        "VpcVPCGWBF912B6E",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "VpcIGWD7BA715C",
        },
        "RouteTableId": Object {
          "Ref": "VpcPublicSubnet1RouteTable6C95E38E",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VpcPublicSubnet1RouteTable6C95E38E": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "http302-demo/Vpc/PublicSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcPublicSubnet1RouteTableAssociation97140677": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VpcPublicSubnet1RouteTable6C95E38E",
        },
        "SubnetId": Object {
          "Ref": "VpcPublicSubnet1Subnet5C2D37C4",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcPublicSubnet1Subnet5C2D37C4": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/20",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "http302-demo/Vpc/PublicSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VpcPublicSubnet2DefaultRoute97F91067": Object {
      "DependsOn": Array [
        "VpcVPCGWBF912B6E",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "VpcIGWD7BA715C",
        },
        "RouteTableId": Object {
          "Ref": "VpcPublicSubnet2RouteTable94F7E489",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VpcPublicSubnet2RouteTable94F7E489": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "http302-demo/Vpc/PublicSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VpcPublicSubnet2RouteTableAssociationDD5762D8": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VpcPublicSubnet2RouteTable94F7E489",
        },
        "SubnetId": Object {
          "Ref": "VpcPublicSubnet2Subnet691E08A3",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VpcPublicSubnet2Subnet691E08A3": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.16.0/20",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          Object {
            "Key": "Name",
            "Value": "http302-demo/Vpc/PublicSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VpcVPCGWBF912B6E": Object {
      "Properties": Object {
        "InternetGatewayId": Object {
          "Ref": "VpcIGWD7BA715C",
        },
        "VpcId": Object {
          "Ref": "Vpc8378EB38",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
    "http30232E1FD3B": Object {
      "Properties": Object {
        "Location": Object {
          "ApplicationId": "arn:aws:serverlessrepo:us-east-1:418289889111:applications/http302-from-origin",
          "SemanticVersion": "1.0.2",
        },
      },
      "Type": "AWS::Serverless::Application",
    },
  },
  "Transform": "AWS::Serverless-2016-10-31",
}
`;
